psi4/driver/p4util/solvers.py

```diff
@@ -310,3 +310,159 @@ def extrapolate(self, out=None):
             out.axpy(c, self.state[num])
 
         return out
+
+
+# Generalized Davidson solver
+
+def davidson_solver(ax_function, preconditioner, guess, e_conv=1.0E-8,
+        r_conv=None, no_eigs=1, max_vecs_per_root=20, maxiter=100):
+    """
+    Solves for the lowest few eigenvalues and eigenvectors of the given real symmetric matrix
+
+    Parameters
+    -----------
+    ax_function : function
+        Takes in a guess vector and returns the Matrix-vector product.
+    preconditioner : function
+        Takes in a list of :py:class:`~psi4.core.Matrix` objects and a mask of active indices. Returns the preconditioned value.
+    guess : list of :py:class:`~psi4.core.Matrix`
+        Starting vectors, required
+    e_conv : float
+        Convergence tolerance for eigenvalues
+    r_conv : float
+        Convergence tolerance for residual vectors
+    no_eigs : int
+        Number of eigenvalues needed
+    maxiter : int
+        The maximum number of iterations this function will take.
+
+    Returns
+    -----------
+
+    Notes
+    -----------
+
+    Examples
+    -----------
+
+    """
+
+    print("Starting Davidson algo:")
+    print("options:")
+    for k,v in dict(e_conv=e_conv, r_conv=r_conv, no_eigs=no_eigs,
+            max_vecs_per_root = max_vecs_per_root, maxiter=maxiter).items():
+        if v == None:
+            v = 'None'
+        print("{:<25}{:.>100}".format(k,v))
+    if r_conv == None:
+        r_conv = e_conv * 100
+    d_tol = 1.0E-8
+
+    # using the shape of the guess vectors to set the dimension of the matrix
+    N = guess.shape[0]
+
+    #sanity check, guess subspace must be at least equal to number of eigenvalues
+    nli = guess.shape[1]
+    if nli < no_eigs:
+        raise ValueError("Not enough guess vectors provided!")
+
+    nl = nli
+    converged=False
+    count = 0
+    A_w_old = np.ones(no_eigs)
+    max_ss_size = no_eigs * max_vecs_per_root
+    B = guess
+
+    ### begin loop
+    conv_roots = [False] * no_eigs
+    while count < maxiter:
+        # active_mask = [True for x in range(nl)] # never used later
+        # Apply QR decomposition on B to orthogonalize the new vectors wrto all other subspace vectors
+        ## orthogonalize preconditioned residuals against all other vectors in the search subspace
+        B, r = np.linalg.qr(B)
+        nl = B.shape[1]
+
+        print("Davidson: Iter={:<3} nl = {:<4}".format(count, nl))
+        # compute sigma vectors corresponding to the new vectors sigma_i = A B_i
+        sigma = np.zeros_like(B)
+        for i in range(nl):
+            sigma[:,i] = ax_function(B[:,i])
```

Comment:
It would be good if you allowed the ax_function to compute all of these vectors at once. Quite a bit of savings can usually be had this way. See the above CG algorithm for similar logic.